sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -567,6 +567,26 @@ object WholeStageCodegenExec {
   def isTooManyFields(conf: SQLConf, dataType: DataType): Boolean = {
     numOfNestedFields(dataType) > conf.wholeStageMaxNumFields
   }
+
+  // The whole stage codegen generates java code on the driver side and sends it to the Executor
+  // side for execution after compilation. The whole codegen can bring significant performance
```

Comment:
`... sends it to the Executors for compilation and execution.`

We also compile code at the driver side, but it's only for checking so that we can fail earlier. The result is discarded.

Reply:
Become enlightened at once. Thank you.